astropy/coordinates/baseframe.py

```diff
@@ -333,39 +331,37 @@ def __init__(
                 f"keywords: {list(kwargs)}"
             )
 
-        # We do ``is None`` because self._data might evaluate to false for
-        # empty arrays or data == 0
-        if self._data is None:
-            # No data: we still need to check that any non-scalar attributes
-            # have consistent shapes. Collect them for all attributes with
-            # size > 1 (which should be array-like and thus have a shape).
-            shapes = {
-                fnm: value.shape
-                for fnm, value in values.items()
-                if getattr(value, "shape", ())
-            }
-            if shapes:
-                if len(shapes) > 1:
-                    try:
-                        self._no_data_shape = check_broadcast(*shapes.values())
-                    except ValueError as err:
-                        raise ValueError(
-                            f"non-scalar attributes with inconsistent shapes: {shapes}"
-                        ) from err
-
-                    # Above, we checked that it is possible to broadcast all
-                    # shapes.  By getting and thus validating the attributes,
-                    # we verify that the attributes can in fact be broadcast.
-                    for fnm in shapes:
-                        getattr(self, fnm)
-                else:
-                    self._no_data_shape = shapes.popitem()[1]
+        # Determine the overall shape of the frame.
+        shapes = [getattr(value, "shape", ()) for value in values.values()]
+        if data is not None:
+            shapes.append(getattr(data, "shape", ()))
+        try:
+            self._shape = check_broadcast(*shapes)
+        except ValueError as err:
+            raise ValueError(
+                f"non-scalar data and/or attributes with inconsistent shapes: {shapes}"
+            ) from err
 
+        # Broadcast the data if necessary and set it
+        if data is not None and getattr(data, "shape", ()) != self._shape:
```

Comment:
I think it should not be possible for `data` not to have a shape, so this can be `data is not None and data.shape != self._shape`.

Reply:
It won't have a shape if it's scalar, will it?

Reply:
Surprisingly, the tests pass if I replace that with `data.shape`. I think that is because the frame data is probably always an Astropy quantity, so it has a shape even if it is scalar. If we ever had a frame that had data that could be a bare Python float, then I think this would break.

Reply:
The frame data should always be a `Representation`, that's why I thought it would always have a shape. I think even the attributes are fine, but am not 100% sure, so probably good to just let those stay the way they are!